mirrord/intproxy/src/proxies/files.rs

```diff
@@ -0,0 +1,1345 @@
+                .await;
+        }
+    }
+
+    #[tracing::instrument(level = Level::DEBUG)]
+    fn protocol_version(&mut self, version: Version) {
+        self.protocol_version.replace(version);
+    }
+
+    // #[tracing::instrument(level = Level::TRACE, skip(message_bus))]
+    async fn file_request(
+        &mut self,
+        request: FileRequest,
+        layer_id: LayerId,
+        message_id: MessageId,
+        message_bus: &mut MessageBus<Self>,
+    ) {
+        match request {
+            // Should trigger remote close only when the fd is closed in all layer instances.
+            FileRequest::Close(close) => {
+                if self.remote_files.remove(layer_id, close.fd) {
+                    self.files_data.remove(&close.fd);
+                    message_bus
+                        .send(ClientMessage::FileRequest(FileRequest::Close(close)))
+                        .await;
+                }
+            }
+
+            // Should trigger remote close only when the fd is closed in all layer instances.
+            FileRequest::CloseDir(close) => {
+                if self.remote_dirs.remove(layer_id, close.remote_fd) {
+                    self.dirs_data.remove(&close.remote_fd);
+                    message_bus
+                        .send(ClientMessage::FileRequest(FileRequest::CloseDir(close)))
+                        .await;
+                }
+            }
+
+            // May require storing additional data in the request queue.
+            FileRequest::Open(open) => {
+                let additional_data = (self.buffer_reads && open.open_options.is_read_only())
+                    .then_some(AdditionalRequestData::OpenBuffered)
+                    .unwrap_or_default();
+                self.request_queue
+                    .insert_with_data(message_id, layer_id, additional_data);
+                message_bus
+                    .send(ClientMessage::FileRequest(FileRequest::Open(open)))
+                    .await;
+            }
+
+            // May require storing additional data in the request queue.
+            FileRequest::OpenRelative(open) => {
+                let additional_data = (self.buffer_reads && open.open_options.is_read_only())
+                    .then_some(AdditionalRequestData::OpenBuffered)
+                    .unwrap_or_default();
+                self.request_queue
+                    .insert_with_data(message_id, layer_id, additional_data);
+                message_bus
+                    .send(ClientMessage::FileRequest(FileRequest::OpenRelative(open)))
+                    .await;
+            }
+
+            // Try to use local buffer if possible.
+            FileRequest::Read(read) => match self.files_data.get_mut(&read.remote_fd) {
+                // File is buffered.
+                Some(data) => {
+                    let from_buffer = data.read_from_buffer(read.buffer_size, data.fd_position);
+                    if let Some(from_buffer) = from_buffer {
```

Comment:
We read from the buffer, if the user asked to read more than what the buffer has, then we go into the `None` case which will re-fill the buffer with new data?

Reply:
Yeah, `None` case will send a `ReadLimited` request to the agent for the next chunk